### PR TITLE
Add more information to the diagnostic bundle

### PR DESCRIPTION
**Pull request by ahoppen**
Add the following information to the diagnose bundle:
- OSLog output of sourcekit-lsp
- Any sourcekit-lsp or SourceKitService crash logs
- The versions of the swift toolchains installed on the system

rdar://123479769

### CODE REVIEW

**Comment by ahoppen** [comment]
@swift-ci Please test

File: Sources/Diagnose/DiagnoseCommand.swift [hunk 1]
```diff
@@ -101,26 +108,130 @@ public struct DiagnoseCommand: AsyncParsableCommand {
...
+  }
+
+  private func addCrashLogs(toBundle bundlePath: URL) throws {
+    #if os(macOS)
+    print("-- Collecting crash reports")
+
+    let destinationDir = bundlePath.appendingPathComponent("crashes")
+    try FileManager.default.createDirectory(at: destinationDir, withIntermediateDirectories: true)
+
+    let diagnosticReports = URL(filePath: "/Library/Logs/DiagnosticReports")
```
Comment (bnbarham) [inline Sources/Diagnose/DiagnoseCommand.swift:153]: Also grab from user? I didn't know these were split!

File: Sources/Diagnose/DiagnoseCommand.swift [hunk 2]
```diff
@@ -101,26 +108,130 @@ public struct DiagnoseCommand: AsyncParsableCommand {
...
+      try process.launch()
+      try await process.waitUntilExit()
+      fileHandle.write("\n".data(using: .utf8)!)
     }
+  }
+
+  public func run() async throws {
+    print(
+      """
+      sourcekit-lsp diagnose collects information that helps developers of sourcekit-lsp diagnose and fix issues.
```
Comment (bnbarham) [inline Sources/Diagnose/DiagnoseCommand.swift:198]:
```suggestion
      sourcekit-lsp diagnose collects information that helps the developers of sourcekit-lsp diagnose and fix issues. 
```

File: Sources/Diagnose/DiagnoseCommand.swift [hunk 3]
```diff
@@ -101,26 +108,130 @@ public struct DiagnoseCommand: AsyncParsableCommand {
...
+    print(
+      """
+      sourcekit-lsp diagnose collects information that helps developers of sourcekit-lsp diagnose and fix issues. 
+      This information contains:
+      - Crash logs from SourceKit
+      - Log messages emitted by SourceKit
+      - Versions of Swift installed on your system
+      - If possible, a minimized project that caused SourceKit to crash
+
+      All information will be collected into a folder on your computer.
```
Comment (bnbarham) [inline Sources/Diagnose/DiagnoseCommand.swift:205]: I'd skip this part and just have `Collection may take a few minutes...`

File: Sources/Diagnose/DiagnoseCommand.swift [hunk 4]
```diff
@@ -101,26 +108,130 @@ public struct DiagnoseCommand: AsyncParsableCommand {
...
+    await orPrintError { try await addSourcekitdCrashReproducer(toBundle: bundlePath) }
+
     print(
       """
-        ----------------------------------------
-        Reduced SourceKit issue and created a bundle that contains a reduced sourcekitd request exhibiting the issue
-        and all the files referenced from the request.
-        The information in this bundle should be sufficient to reproduce the issue.
+      ----------------------------------------
+      Created a bundle that contains information about sourcekit-lsp issues encountered recently.
```
Comment (bnbarham) [inline Sources/Diagnose/DiagnoseCommand.swift:225]: `Bundle created. When filing an issue at https://github.com/apple/sourcekit-lsp/issues/new, please attach the bundle located at \(bundlePath.path).`

File: Sources/Diagnose/DiagnoseCommand.swift [hunk 5]
```diff
@@ -101,26 +108,130 @@ public struct DiagnoseCommand: AsyncParsableCommand {
...
+      try await process.waitUntilExit()
+      fileHandle.write("\n".data(using: .utf8)!)
     }
+  }
+
+  public func run() async throws {
+    print(
+      """
+      sourcekit-lsp diagnose collects information that helps developers of sourcekit-lsp diagnose and fix issues. 
+      This information contains:
```
Comment (bnbarham) [inline Sources/Diagnose/DiagnoseCommand.swift:199]: Join to previous and make `This includes:`?

File: Sources/Diagnose/DiagnoseCommand.swift [hunk 6]
```diff
@@ -101,26 +108,130 @@ public struct DiagnoseCommand: AsyncParsableCommand {
...
+    print("-- Collecting crash reports")
+
+    let destinationDir = bundlePath.appendingPathComponent("crashes")
+    try FileManager.default.createDirectory(at: destinationDir, withIntermediateDirectories: true)
+
+    let diagnosticReports = URL(filePath: "/Library/Logs/DiagnosticReports")
+    let enumerator = FileManager.default.enumerator(at: diagnosticReports, includingPropertiesForKeys: nil)
+    while let fileUrl = enumerator?.nextObject() as? URL {
+      guard
+        fileUrl.lastPathComponent.hasPrefix("SourceKitService") || fileUrl.lastPathComponent.hasPrefix("sourcekit-lsp")
```
[the remark this answers is not in the document]
Comment (ahoppen) [reply Sources/Diagnose/DiagnoseCommand.swift:157]: Good idea 👍🏽

File: Sources/Diagnose/DiagnoseCommand.swift (hunk 5 above)
Comment (ahoppen) [reply Sources/Diagnose/DiagnoseCommand.swift:199]: Which previous do you mean?

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by ahoppen** [comment]
@swift-ci Please test Windows

**Pull request merged**